### PR TITLE
ci, build: update CI, libraries, python dependencies, and format

### LINKED ISSUE #141: Unit tests are failing with assertion fails
Some test cases are failing due to:

```
======================================================================================================== short test summary info ========================================================================================================
FAILED tests/unit/test_operator.py::TestCharm::test_create_resources_success - AssertionError: Expected 'apply' to have been called once. Called 0 times.
FAILED tests/unit/test_operator.py::TestCharm::test_main - AssertionError: Expected 'apply' to have been called.
FAILED tests/unit/test_sidebar_lib.py::TestDashboardLink::test_location_validation[invalid-location-context_raised4] - Failed: DID NOT RAISE <class 'ValueError'>
===================================================================================================== 3 failed, 39 passed in 0.45s ======================================================================================================
ERROR: InvocationError for command /home/ubuntu/kubeflow-dashboard-operator/.tox/unit/bin/coverage run --source=/home/ubuntu/kubeflow-dashboard-operator/src/ -m pytest --ignore=/home/ubuntu/kubeflow-dashboard-operator/tests/integration -vv --tb native (exited with code 1)
________________________________________________________________________________________________________________ summary ________________________________________________________________________________________________________________
ERROR:   unit: commands failed
```

Latest run in #140 is an example of this.

### LINKED ISSUE #147: `kubeflow-dashboard-operator` fails to build during integration tests
Seems like this charm cannot be built during integration tests, resulting in the following error message:

```
RuntimeError: Failed to build charm .:
Packing the charm.
Launching environment to pack for base name='ubuntu' channel='20.04' architectures=['amd64'] (may take a while the first time but it's reusable)
Packing the charm
Packing the charm.
Building charm in '/root'
Running step PULL for part 'charm'
Running step BUILD for part 'charm'
Parts processing error: Failed to run the build script for part 'charm'.
Failed to build charm for bases index '0'.
Full execution log: '/home/runner/.local/state/charmcraft/log/charmcraft-20230724-102034.513526.log'
```

Logs and details in the latest [integration tests job](https://github.com/canonical/kubeflow-dashboard-operator/actions/runs/5643586071/job/15285722021) for #140

### PR DESCRIPTION
**Pull request by DnPlas**
This commit updates the CI with standard CI configuration (juju, actions, deps) that is shared across multiple repositories. All python dependencies are also updated to their latest working versions. This PR also includes library updates. Finally, some formatting was applied on non-compliant files (copyright, style).

This change also contains a fix for #147, an issue caused by a missing toolchain in the charm that allows building certain python packages. For more information, please refer to canonical/bundle-kubeflow#648.

Fixes #141
Fixes #147
Part of canonical/bundle-kubeflow#648

### CODE REVIEW

File: charmcraft.yaml [hunk 1]
```diff
@@ -13,5 +13,4 @@ parts:
   charm:
     prime:
       - files/*yaml
```
Comment (phoevos) [inline charmcraft.yaml:15]: Should we also remove this? `files` does not exist anymore and `templates` gets pushed automatically.

File: requirements.in
Comment (phoevos) [inline requirements.in:1]: Let's add `pyyaml` here, needed by `dashboard_links.py`.

File: requirements-integration.in
Comment (phoevos) [inline requirements-integration.in:1]: `charmed-kubeflow-chisme` and `serialized-data-interface` are not needed in the integration tests. I would add `pytest` and `pytest-asyncio` though, which are directly imported.

File: tests/unit/test_operator.py [hunk 2]
```diff
@@ -193,6 +193,8 @@ def test_create_resources_success(
         harness_with_profiles: Harness,
     ):
         harness_with_profiles.begin()
+        container = harness_with_profiles.charm.model.unit.get_container(CHARM_NAME)
+        harness_with_profiles.set_can_connect(container, True)
```
Comment (phoevos) [inline tests/unit/test_operator.py:197]: Brilliant, I wouldn't change a thing.

**Review by phoevos — CHANGES_REQUESTED** [review]

File: charmcraft.yaml (hunk 1 above)
Comment (DnPlas) [reply charmcraft.yaml:15]: This sounds like a candidate for a separate PR, I can make the change, but I wouldn't want this to get buried in the squash of this commit and gets lost if someone wants to check when this change was made.

File: requirements-integration.in
Comment (phoevos) [reply requirements-integration.in:1]: > `charmed-kubeflow-chisme` and `serialized-data-interface` are not needed in the integration tests.

Well, I've been wrong before.

**Review by phoevos — APPROVED** [review]

**Pull request merged**